src/openzeppelin/tests/test_erc721.cairo

```diff
@@ -0,0 +1,475 @@
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: approval to owner', ))]
+fn test__approve_to_owner() {
+    setup();
+
+    ERC721::_approve(OWNER(), TOKEN_ID());
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: invalid token ID', ))]
+fn test__approve_nonexistent() {
+    ERC721::_approve(SPENDER(), TOKEN_ID());
+}
+
+///
+/// set_approval_for_all & _set_approval_for_all
+///
+
+#[test]
+#[available_gas(2000000)]
+fn test_set_approval_for_all_true() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OPERATOR(), true);
+    assert(ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test_set_approval_for_all_false() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OPERATOR(), false);
+    assert(!ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test_set_approval_for_all_owner_equal_operator_true() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OWNER(), true);
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test_set_approval_for_all_owner_equal_operator_false() {
+    set_caller_address(OWNER());
+    ERC721::set_approval_for_all(OWNER(), false);
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test__set_approval_for_all_true() {
+    ERC721::_set_approval_for_all(OWNER(), OPERATOR(), true);
+    assert(ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+fn test__set_approval_for_all_false() {
+    ERC721::_set_approval_for_all(OWNER(), OPERATOR(), false);
+    assert(!ERC721::is_approved_for_all(OWNER(), OPERATOR()), 'Operator not approved correctly');
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test__set_approval_for_all_owner_equal_operator_true() {
+    ERC721::_set_approval_for_all(OWNER(), OWNER(), true);
+}
+
+#[test]
+#[available_gas(2000000)]
+#[should_panic(expected = ('ERC721: self approval', ))]
+fn test__set_approval_for_all_owner_equal_operator_false() {
+    ERC721::_set_approval_for_all(OWNER(), OWNER(), false);
+}
+
+///
+/// transfer_from
+///
+
+#[test]
+#[available_gas(2000000)]
+fn test_transfer_from_owner() {
+    setup();
+
+    assert(ERC721::owner_of(TOKEN_ID()) == OWNER(), 'Ownership before');
+    assert(ERC721::balance_of(OWNER()) == u256_from_felt252(1), 'Balance of owner before');
+    assert(ERC721::balance_of(RECIPIENT()) == u256_from_felt252(0), 'Balance of recipient before');
+
+    set_caller_address(OWNER());
+    ERC721::transfer_from(OWNER(), RECIPIENT(), TOKEN_ID());
+
+    assert(ERC721::owner_of(TOKEN_ID()) == RECIPIENT(), 'Ownership after');
```

Comment:
should we add `is_approved_for_all` to the checks?

Reply:
Approval for all is not attached to a token in particular. Which combo (owner/spender) would you test?

IMO that kind of check should be done in FV, because what we would like to check is:
forall user, forall spender, `is_approval_for_all` should return the same value before and after a call to `ERC721::transfer_from`